src/chainstate/burn/operations/peg_in.rs

```diff
@@ -0,0 +1,451 @@
+use clarity::codec::StacksMessageCodec;
+use clarity::vm::errors::RuntimeErrorType as ClarityRuntimeError;
+use clarity::vm::types::PrincipalData;
+use clarity::vm::types::QualifiedContractIdentifier;
+use clarity::vm::types::StandardPrincipalData;
+use clarity::vm::ContractName;
+
+use crate::burnchains::BurnchainBlockHeader;
+use crate::burnchains::BurnchainTransaction;
+use crate::chainstate::burn::Opcodes;
+use crate::types::chainstate::StacksAddress;
+use crate::types::Address;
+
+use crate::chainstate::burn::operations::Error as OpError;
+use crate::chainstate::burn::operations::PegInOp;
+
+impl PegInOp {
+    pub fn from_tx(
+        block_header: &BurnchainBlockHeader,
+        tx: &BurnchainTransaction,
+    ) -> Result<Self, OpError> {
+        if tx.opcode() != Opcodes::PegIn as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(OpError::InvalidInput);
+        }
+
+        let (amount, peg_wallet_address) =
+            if let Some(Some(recipient)) = tx.get_recipients().first() {
+                (recipient.amount, recipient.address.clone())
+            } else {
+                warn!("Invalid tx: Output 2 not provided");
+                return Err(OpError::InvalidInput);
+            };
+
+        let recipient = Self::parse_data(&tx.data())?;
+
+        let txid = tx.txid();
+        let vtxindex = tx.vtxindex();
+        let block_height = block_header.block_height;
+        let burn_header_hash = block_header.block_hash;
+
+        Ok(Self {
+            recipient,
+            peg_wallet_address,
+            amount,
+            txid,
+            vtxindex,
+            block_height,
+            burn_header_hash,
+        })
+    }
+
+    fn parse_data(data: &[u8]) -> Result<PrincipalData, ParseError> {
+        /*
+            Wire format:
+
+            0      2  3                  24                            64
+            |------|--|------------------|-----------------------------|
+             magic  op   Stacks address      Contract name (optional)
+
+             Note that `data` is missing the first 3 bytes -- the magic and op must
+             be stripped before this method is called. At the time of writing,
+             this is done in `burnchains::bitcoin::blocks::BitcoinBlockParser::parse_data`.
+        */
+
+        if data.len() < 21 {
+            warn!(
+                "PegInOp payload is malformed ({} bytes, expected at least {})",
+                data.len(),
+                21
+            );
+            return Err(ParseError::MalformedData);
+        }
+
+        let version = *data.get(0).ok_or(ParseError::MalformedData)?;
+        let address_data: [u8; 20] = data
+            .get(1..21)
+            .ok_or(ParseError::MalformedData)?
+            .try_into()?;
+
+        let standard_principal_data = StandardPrincipalData(version, address_data);
+
+        let maybe_contract_name: Option<ContractName> = data
```

Comment:
This code does not handle the case if there are trailing 0-bytes, or even the case where the `name` field is all 0's.  I think it should, because the space beyond the 64th byte can be used as a memo field for future extensions.  Instead, this code should just select out the leading non-zero bytes from the range 21 to 61, and try to decode _that_ into a ContractName.  If there are trailing 0's, it's fine -- they'll be ignored.  If the whole name field is 0, then there's no name.

Also, please don't use such long operator chains!  It's very hard to follow what this code is supposed to be doing.  Just do something like this:

```rust
let maybe_contract_name: Option<ContractName> = {
   // read to the first 0 byte
   let strlen = {
      let mut len = cmp::min(61, data.len());
      for i in 21..len {
         if data[i] == 0 {
            len = i;
            break;
          }
       }
       len
   };
      
   let name_bytes = data.get(21..strlen).unwrap_or(&[]);
   let name_string = std::str::from_utf8(&name_bytes).map_err(|_| ParseError::MalformedData)?.to_string();
   let name = ContractName::try_from(name_str).map_err(|_| ParseError::MalformedData)?;
   name
}
```

Reply:
Alright. I factored out some pieces of the suggestion to a `leading_non_zero_bytes` utility function and updated the functionality according to your description. It should be more readable now without the long operator chain. Let me know if you have any further opinions.